# Web Scraping

## Top Data Analytics Companies
To improve effectiveness in business processes, companies are focussing on collecting and utilizing data. Data analytics companies enable businesses to analyze the acquired data and use them as required. Data analytics services can assist in product development, identifying potential market gaps, improving operational efficiency, etc. [Goodfirms](https://www.goodfirms.co/big-data-analytics/data-analytics)

The aim of this project is to scrape Goodfirms website for details about top data analytics company. These details includes, review, rating, year founded, location, etc.

### Road Map
Stage 1
* Download page
* Open page
* Get details
* Extract details
* Make a dataframe

Stage 2 (Code Refactoring)
* Change loops to function
* Take functions to another file
* Import functions and extract
* Make a dataframe

### Libraries

In [1]:
import requests as r
from bs4 import BeautifulSoup
import pandas as pd
import plotly.express as px
import plotly.offline as po
po.init_notebook_mode(connected=True)

### Download and save HTML file

In [5]:
# URL link
url_link = ['https://www.goodfirms.co/big-data-analytics/data-analytics', 'https://www.goodfirms.co/big-data-analytics/data-analytics?page=2', 
       'https://www.goodfirms.co/big-data-analytics/data-analytics?page=3']

# access website
access = [r.get(url) for url in url_link]

""" OR

access = []
for url in url_link:
    access.append(r.get(url))
    
"""


### Saving webpage to PC

In [15]:
for page in access:
    index = access.index(page)+1
    with open(f'page{index}.html', mode='wb') as file:
        file.write(page.content)

### Page 1 

In [34]:
with open("page1.html", encoding='utf-8', mode='r') as file:
    bs = BeautifulSoup(file, 'lxml')

#### Locate Details

In [35]:
firm_names = bs.find_all('span', {'itemprop': "name"})
firm_motors = bs.find_all('p', {'class': "profile-tagline"})
firm_reviews = bs.find_all('span', {'class': "listinv_review_label"})
progress_value = bs.find_all('div', {'class': "circle-progress-value"})
firm_prices = bs.find_all('div', {'class': "firm-pricing"})
firm_emps = bs.find_all('div', {'class': "firm-employees"})
firm_years = bs.find_all('div', {'class': "firm-founded"})
firm_locations = bs.find_all('div', {'class': "firm-location"})

#### Function to extract details

In [36]:
def extract_detail(tag_lst):
    lst = [tag.text for tag in tag_lst]
    return pd.Series(lst)

def extract_progress_values(tag_lst):    
    service_pct = [percent[1].text for percent in enumerate(tag_lst) if percent[0]%2 ==0]
    platform_pct = [percent[1].text for percent in enumerate(tag_lst) if percent[0]%2 ==1]       
    return pd.Series(service_pct), pd.Series(platform_pct) 

#### Extract Details

In [37]:
names = extract_detail(firm_names[3:])
motors = extract_detail(firm_motors)
reviews = extract_detail(firm_reviews)
ser, pct = extract_progress_values(progress_value)
prices = extract_detail(firm_prices)
emps = extract_detail(firm_emps)
years = extract_detail(firm_years)
locations = extract_detail(firm_locations)

### Put Details in a DataFrame

In [38]:
df1 = pd.DataFrame()

In [39]:
df1['firm_name'] = names
df1['firm_motor'] = motors
df1['firm_review'] = reviews
df1['service_pct'] = ser
df1['platform_pct'] = pct
df1['firm_price'] = prices
df1['firm_employee'] = emps
df1['year_founded'] = years
df1['firm_location'] = locations

In [40]:
df1.head()

firm_name                                    firm_motor  \
0          SPEC INDIA  Enterprise Software, Mobility & BI Solutions   
1        instinctools                   Delivering the future. Now.   
2             SoluLab   Blockchain | IoT | Mobility | AI | Big Data   
3  Sigma Data Systems       Discover the world of Big Data with us!   
4       NeenOpal Inc.            The Hub Of Data Science Innovation   

        firm_review service_pct platform_pct       firm_price firm_employee  \
0  4.8 (26 Reviews)         20%          15%      \n< $25/hr      250 - 999   
1   4.8 (8 Reviews)          5%          10%  \n$50 - $99/hr      250 - 999   
2  5.0 (32 Reviews)         15%          15%  \n$25 - $49/hr       50 - 249   
3   4.7 (5 Reviews)         40%          10%  \n$25 - $49/hr      250 - 999   
4   5.0 (5 Reviews)         25%          10%  \n$25 - $49/hr        10 - 49   

  year_founded                firm_location  
0         1987      \nIndia, United States   
1         2000    \nUnited States, Germany   
2         2014      \nUnited States, India   
3         2010  \nUnited States, Australia   
4         2016     \nUnited States, India

Page 2

In [41]:
# Open page 2
with open("page2.html", encoding='utf-8', mode='r') as file:
    bs2 = BeautifulSoup(file, 'lxml')
    
# Locate details
firm_names = bs2.find_all('span', {'itemprop': "name"})
firm_motors = bs2.find_all('p', {'class': "profile-tagline"})
firm_reviews = bs2.find_all('span', {'class': "listinv_review_label"})
progress_value = bs2.find_all('div', {'class': "circle-progress-value"})
firm_prices = bs2.find_all('div', {'class': "firm-pricing"})
firm_emps = bs2.find_all('div', {'class': "firm-employees"})
firm_years = bs2.find_all('div', {'class': "firm-founded"})
firm_locations = bs2.find_all('div', {'class': "firm-location"})

# Extract details
names2 = extract_detail(firm_names[3:])
motors = extract_detail(firm_motors)
reviews = extract_detail(firm_reviews)
ser, pct = extract_progress_values(progress_value)
prices = extract_detail(firm_prices)
emps = extract_detail(firm_emps)
years = extract_detail(firm_years)
locations = extract_detail(firm_locations)

# Put details in a dataframe
df2 = pd.DataFrame()
df2['firm_name'] = names2
df2['firm_motor'] = motors
df2['firm_review'] = reviews
df2['service_pct'] = ser
df2['platform_pct'] = pct
df2['firm_price'] = prices
df2['firm_employee'] = emps
df2['year_founded'] = years
df2['firm_location'] = locations

# Preview
df2.head()

firm_name  \
0                                       Inoxoft   
1  Noventum Custom Software Development Company   
2                 Napollo Software Design L.L.C   
3                      SemiDot Infotech Pvt Ltd   
4                              Sphinx Solutions   

                                          firm_motor      firm_review  \
0  IEC/ISO 27001, Google, Microsoft Certified Com...  5.0 (8 Reviews)   
1    Custom Software & Web Development in New Mexico  5.0 (3 Reviews)   
2    Best Software Design Agency in Dubai & New York  5.0 (7 Reviews)   
3          Right Technology Partner for IT Solutions  4.7 (7 Reviews)   
4                        Inspire : Innovate : Evolve  4.6 (8 Reviews)   

  service_pct platform_pct         firm_price firm_employee year_founded  \
0         10%          10%    \n$25 - $49/hr       50 - 249         2014   
1         20%         100%  \n$100 - $149/hr          2 - 9         2012   
2         20%          20%              \nNA       50 - 249         2011   
3         10%          30%        \n< $25/hr       50 - 249         2011   
4          5%          10%        \n< $25/hr     Freelancer         2010   

             firm_location  
0               \nUkraine   
1         \nUnited States   
2         \nUnited States   
3  \nUnited States, India   
4  \nIndia, United States

Page 3

In [42]:
# Open page 3 
with open("page3.html", encoding='utf-8', mode='r') as file:
    bs3 = BeautifulSoup(file, 'lxml')
    
# Locate details
firm_names = bs3.find_all('span', {'itemprop': "name"})
firm_motors = bs3.find_all('p', {'class': "profile-tagline"})
firm_reviews = bs3.find_all('span', {'class': "listinv_review_label"})
progress_value = bs3.find_all('div', {'class': "circle-progress-value"})
firm_prices = bs3.find_all('div', {'class': "firm-pricing"})
firm_emps = bs3.find_all('div', {'class': "firm-employees"})
firm_years = bs3.find_all('div', {'class': "firm-founded"})
firm_locations = bs3.find_all('div', {'class': "firm-location"})

# Extract details
names = extract_detail(firm_names[3:])
motors = extract_detail(firm_motors)
reviews = extract_detail(firm_reviews)
ser, pct = extract_progress_values(progress_value)
prices = extract_detail(firm_prices)
emps = extract_detail(firm_emps)
years = extract_detail(firm_years)
locations = extract_detail(firm_locations)

# Put details in a dataframe
df3 = pd.DataFrame()
df3['firm_name'] = names
df3['firm_motor'] = motors
df3['firm_review'] = reviews
df3['service_pct'] = ser
df3['platform_pct'] = pct
df3['firm_price'] = prices
df3['firm_employee'] = emps
df3['year_founded'] = years
df3['firm_location'] = locations

# Preview
df3.head()

firm_name  \
0  Techmango Technology Services Private Limited   
1                            Evolve Technologies   
2                    Virtual Electronics PTE LTD   
3                                        Reenbit   
4                         Michigan Software Labs   

                                         firm_motor      firm_review  \
0       Best Offshore Software Development Company   4.3 (3 Reviews)   
1                                         IT agency   4.8 (1 Review)   
2  Software and mobile app development in Singapore  5.0 (2 Reviews)   
3                 Intelligent engineering & beyond   5.0 (3 Reviews)   
4                                      We Make Apps  5.0 (2 Reviews)   

  service_pct platform_pct         firm_price firm_employee year_founded  \
0         10%          20%        \n< $25/hr      250 - 999         2014   
1         30%          30%    \n$50 - $99/hr        10 - 49         2000   
2          5%          25%    \n$25 - $49/hr        10 - 49         2019   
3          5%           5%    \n$25 - $49/hr       50 - 249         2018   
4         10%          20%  \n$150 - $199/hr        10 - 49         2010   

        firm_location  
0            \nIndia   
1          \nIreland   
2        \nSingapore   
3  \nUkraine, Poland   
4    \nUnited States

### Combine all DataFrame

In [44]:
all_df = pd.concat([df1, df2, df3], ignore_index=True)
all_df

firm_name                                         firm_motor  \
0            SPEC INDIA       Enterprise Software, Mobility & BI Solutions   
1          instinctools                        Delivering the future. Now.   
2               SoluLab        Blockchain | IoT | Mobility | AI | Big Data   
3    Sigma Data Systems            Discover the world of Big Data with us!   
4         NeenOpal Inc.                 The Hub Of Data Science Innovation   
..                  ...                                                ...   
148      Rock Your Data    Delivering Cloud Analytics that Rocks Your Data   
149          CodeRiders                    We desire. Together we achieve!   
150          Notionmind              Software. Strategy. Managed Services.   
151          OptimusFox         Best Blockchain Development Company in USA   
152          QAT Global  Together We Can Accomplish Anything - US - BR- CR   

          firm_review service_pct platform_pct         firm_price  \
0    4.8 (26 Reviews)         20%          15%        \n< $25/hr    
1     4.8 (8 Reviews)          5%          10%    \n$50 - $99/hr    
2    5.0 (32 Reviews)         15%          15%    \n$25 - $49/hr    
3     4.7 (5 Reviews)         40%          10%    \n$25 - $49/hr    
4     5.0 (5 Reviews)         25%          10%    \n$25 - $49/hr    
..                ...         ...          ...                ...   
148    0.0 (0 Review)         70%          10%  \n$150 - $199/hr    
149    0.0 (0 Review)         15%          50%    \n$25 - $49/hr    
150    0.0 (0 Review)         15%          50%   \n$50 - $99/hr     
151    0.0 (0 Review)         15%          50%    \n$50 - $99/hr    
152    0.0 (0 Review)         10%          50%    \n$50 - $99/hr    

    firm_employee year_founded                firm_location  
0       250 - 999         1987      \nIndia, United States   
1       250 - 999         2000    \nUnited States, Germany   
2        50 - 249         2014      \nUnited States, India   
3       250 - 999         2010  \nUnited States, Australia   
4         10 - 49         2016     \nUnited States, India    
..            ...          ...                          ...  
148       10 - 49         2017                    \nCanada   
149       10 - 49         2013                  \n Armenia   
150       10 - 49         2019                     \nIndia   
151      50 - 249         2018             \nUnited States   
152      50 - 249         1995     \nUnited States, Brazil   

[153 rows x 9 columns]

### Data Cleaning

In [45]:
df = all_df.copy()
df.head(5)

firm_name                                    firm_motor  \
0          SPEC INDIA  Enterprise Software, Mobility & BI Solutions   
1        instinctools                   Delivering the future. Now.   
2             SoluLab   Blockchain | IoT | Mobility | AI | Big Data   
3  Sigma Data Systems       Discover the world of Big Data with us!   
4       NeenOpal Inc.            The Hub Of Data Science Innovation   

        firm_review service_pct platform_pct       firm_price firm_employee  \
0  4.8 (26 Reviews)         20%          15%      \n< $25/hr      250 - 999   
1   4.8 (8 Reviews)          5%          10%  \n$50 - $99/hr      250 - 999   
2  5.0 (32 Reviews)         15%          15%  \n$25 - $49/hr       50 - 249   
3   4.7 (5 Reviews)         40%          10%  \n$25 - $49/hr      250 - 999   
4   5.0 (5 Reviews)         25%          10%  \n$25 - $49/hr        10 - 49   

  year_founded                firm_location  
0         1987      \nIndia, United States   
1         2000    \nUnited States, Germany   
2         2014      \nUnited States, India   
3         2010  \nUnited States, Australia   
4         2016     \nUnited States, India

In [46]:
df.insert(2, 'star_rating', df['firm_review'].apply(lambda x: x.split()[0]))

In [47]:
df.insert(3, 'review', df['firm_review'].apply(lambda x: x.split()[1].strip('(')))

In [48]:
df.rename(columns={'service_pct':'service_pct(%)'}, inplace=True)
df['service_pct(%)'] = df['service_pct(%)'].apply(lambda x: x.strip('%'))

In [49]:
df.rename(columns={'platform_pct':'platform_pct(%)'}, inplace=True)
df['platform_pct(%)'] = df['platform_pct(%)'].apply(lambda x: x.strip('%'))

In [50]:
df['firm_price'] = df['firm_price'].apply(lambda x: x.strip('\n'))

In [55]:
df['firm_location'] = df['firm_location'].apply(lambda x: x.strip('\n'))

In [57]:
df.drop(columns='firm_review', inplace=True)

In [58]:
df.head(5)

firm_name                                    firm_motor  \
0          SPEC INDIA  Enterprise Software, Mobility & BI Solutions   
1        instinctools                   Delivering the future. Now.   
2             SoluLab   Blockchain | IoT | Mobility | AI | Big Data   
3  Sigma Data Systems       Discover the world of Big Data with us!   
4       NeenOpal Inc.            The Hub Of Data Science Innovation   

  star_rating review service_pct(%) platform_pct(%)     firm_price  \
0         4.8     26             20              15      < $25/hr    
1         4.8      8              5              10  $50 - $99/hr    
2         5.0     32             15              15  $25 - $49/hr    
3         4.7      5             40              10  $25 - $49/hr    
4         5.0      5             25              10  $25 - $49/hr    

  firm_employee year_founded              firm_location  
0     250 - 999         1987      India, United States   
1     250 - 999         2000    United States, Germany   
2      50 - 249         2014      United States, India   
3     250 - 999         2010  United States, Australia   
4       10 - 49         2016     United States, India

In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   firm_name        153 non-null    object 
 1   firm_motor       153 non-null    object 
 2   star_rating      153 non-null    float64
 3   review           153 non-null    int32  
 4   service_pct(%)   153 non-null    int32  
 5   platform_pct(%)  153 non-null    int32  
 6   firm_price       153 non-null    object 
 7   firm_employee    153 non-null    object 
 8   year_founded     153 non-null    object 
 9   firm_location    153 non-null    object 
dtypes: float64(1), int32(3), object(6)
memory usage: 10.3+ KB


In [61]:
df['star_rating'] = df['star_rating'].astype('float')
df['review'] = df['review'].astype('int')
df['service_pct(%)'] = df['service_pct(%)'].astype('int')
df['platform_pct(%)'] = df['platform_pct(%)'].astype('int')

### EDA

In [63]:
top5_sta_rev = df.sort_values(by=['star_rating', 'review'], ascending=False)[:5]
top5_sta_rev

firm_name                                   firm_motor  star_rating  \
2     SoluLab  Blockchain | IoT | Mobility | AI | Big Data          5.0   
31      S-PRO                           Strategic partner.          5.0   
35    MobiDev          Web and Mobile development, AI apps          5.0   
39   APPSTIRR        Award Winning App Development Company          5.0   
41  Idealogic       End-to-end custom software development          5.0   

    review  service_pct(%)  platform_pct(%)     firm_price firm_employee  \
2       32              15               15  $25 - $49/hr       50 - 249   
31      30              10               25  $50 - $99/hr       50 - 249   
35      16              10               10  $50 - $99/hr      250 - 999   
39      14               5               20  $25 - $49/hr       50 - 249   
41      13              10               20  $50 - $99/hr       50 - 249   

   year_founded                         firm_location  
2          2014                 United States, India   
31         2014                               Poland   
35         2009                        United States   
39         2010  United States, United Arab Emirates   
41         2016                     Ukraine, Estonia

In [64]:
px.histogram(df, 'star_rating', width=500, title='Star Rating Distribution')

In [ ]:
px.histogram(df, 'review', width=500, title='Review Distribution')

In [ ]:
px.bar(top5_sta_rev, 'firm_name', ['star_rating', 'review'], width=700)

In [ ]:
top5_sta_ser = df.sort_values(by=['star_rating', 'service_pct(%)'], ascending=False)[:5]
top5_sta_ser

firm_position    firm_name  star_rating  review  \
8               9     Datapine          5.0       3   
24             25     SetuServ          5.0       1   
9              10   Dataforest          5.0       4   
20             21       VITech          5.0       3   
25             26  InData Labs          5.0       1   

                                       firm_motor  service_pct(%)  \
8                 BUSINESS INTELLIGENCE MADE EASY              90   
24  Customized AI: Machine Power + Human Accuracy              50   
9    Data Engineering and Web Product Development              30   
20                Software development powerhouse              30   
25                     Big on Data Science and AI              30   

    platform_pct(%)      firm_price firm_employee year_founded  \
8                 5   $50 - $99/hr        10 - 49         2012   
24               16             NA       50 - 249         2012   
9                20   $50 - $99/hr       50 - 249         2018   
20               25  $50 - $99/hr        50 - 249         2007   
25               30   $50 - $99/hr       50 - 249         2014   

            firm_location  
8                Germany   
24  United States, India   
9       Ukraine, Estonia   
20               Ukraine   
25     Cyprus, Singapore

In [ ]:
px.bar(top5_sta_ser, 'firm_name', ['star_rating', 'service_pct(%)'], width=600)

In [ ]:
df.sort_values('year_founded')[:5]

firm_position            firm_name  star_rating  review  \
0               1           SPEC INDIA          4.8      26   
1               2         instinctools          4.8       8   
33             34  CDN Solutions Group          4.6      20   
46             47           HQSoftware          4.9      11   
17             18            OrbitSoft          4.9       3   

                                           firm_motor  service_pct(%)  \
0        Enterprise Software, Mobility & BI Solutions              20   
1                         Delivering the future. Now.               5   
33  Blockchain, NFT, Mobile App, Software Development              10   
46       IoT, AR, VR, and custom software development               5   
17                               We do software right               5   

    platform_pct(%)     firm_price firm_employee year_founded  \
0                15      < $25/hr      250 - 999         1987   
1                10  $50 - $99/hr      250 - 999         2000   
33               20      < $25/hr      250 - 999         2000   
46               25  $25 - $49/hr       50 - 249         2001   
17               20  $25 - $49/hr       50 - 249         2001   

              firm_location  
0     India, United States   
1   United States, Germany   
33    India, United States   
46  United States, Georgia   
17           United States

In [ ]:
df['firm_price'].value_counts()

$25 - $49/hr       27
$50 - $99/hr       11
< $25/hr            6
$50 - $99/hr        3
NA                  3
$100 - $149/hr      1
Name: firm_price, dtype: int64

In [ ]:
split = df['firm_location'].apply(lambda x: x.split(', '))
country_lst = []
for lst in split:
    country_lst += lst
new = []
for country in country_lst:
    new.append(country.strip())    

countries_cnt = pd.Series(new).value_counts(ascending=True)

In [ ]:
px.bar(x=countries_cnt.values, y=countries_cnt.index, labels={'x': 'count', 'y':''})

### Summary

* The webpage from Goodfirm was succesfully accessed and downloaded using Request library. 
* BeautifulSoup was used to locate and extract the details from the downloaded html file
* The extracted details was converted to data frame using Pandas.
* The file was cleaned and changed to its right datatype
* The cleaned file was explored for some insights.

#### Findings
* About 30 firms are rated 5 star while 12 are rated between 4.8 - 4.9
* Very few firms (3) has a review above 30
* Most of the firms are located in United States followed by India.